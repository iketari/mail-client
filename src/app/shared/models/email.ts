export interface IEmail {
    from: string;
    to: string[];
    cc: string[];
    bcc: string[];
    subject: string;
    body: string;
    date: string;
    id?: string;
}