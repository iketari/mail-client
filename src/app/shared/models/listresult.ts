export interface IListResult<T> {
  items: T[];
  page: number;
  limit: number;
  total: number;
}
