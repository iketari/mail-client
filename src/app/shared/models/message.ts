export interface IEmail {
  from: string;
  to: string[];
  cc: string[];
  bcc: string[];
  subject: string;
  body: string;
  date: Date;
  id?: string;
}

export interface IThread<T> {
  messages: T[];
  subject: string;
  participants: string[];
  id: string;
}
