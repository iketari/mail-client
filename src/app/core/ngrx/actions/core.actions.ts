import { Action } from '@ngrx/store';
import { IEmail } from '../../../shared/models/message';
import { IListResult } from '../../../shared/models/listresult';

export enum CoreActionTypes {
  LoadEmails = '[Core] Load Emails',
  LoadEmailsSuccess = '[Core] Load Emails Success',
  LoadEmailsFail = '[Core] Load Emails Fail'
}

export class LoadEmails implements Action {
  readonly type = CoreActionTypes.LoadEmails;
  constructor(public payload: { page: number }) {}
}

export class LoadEmailsSuccess implements Action {
  readonly type = CoreActionTypes.LoadEmailsSuccess;
  constructor(public payload: IListResult<IEmail>) {}
}

export class LoadEmailsFail implements Action {
  readonly type = CoreActionTypes.LoadEmailsFail;
  constructor(public payload: any) {}
}

export type CoreActions = LoadEmails | LoadEmailsSuccess | LoadEmailsFail;
