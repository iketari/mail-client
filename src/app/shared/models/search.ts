import { IListResult } from './listresult';
import { IThread } from './message';

export interface ISearchQuery {
  date_from: Date;
  date_to: Date;
  query: string;
  from: string;
  to: string[];
}

export interface ISearchResult<T> {
  originalItem: T;
  filteredBy: {
    [field: string]: boolean;
  };
  highlights: {
    [field: string]: THighlightRange[];
  };
}

export interface IThreadSearchResult<T> extends IThread<T> {
  performedMessages: ISearchResult<T>[];
  highlights: {
    [field: string]: THighlightRange[];
  };
}

export interface ISearchResponse<T> {
  items: IThreadSearchResult<T>[];
  page: number;
  limit: number;
  total: number;
  context: ISearchQuery;
  participants: IParticipant[];
}

export type THighlightRange = [number, number];

export interface IParticipant {
  id?: string;
  email: string;
  to: IParticipant[];
  toEntities: {
    [key: string]: IParticipant;
  };
}
