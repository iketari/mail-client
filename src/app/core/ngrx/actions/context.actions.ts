import { Action } from '@ngrx/store';
import { IEmail } from '../../../shared/models/message';
import { IListResult } from '../../../shared/models/listresult';
import { IParticipant } from '../../../shared/models/search';

export enum ContextActionTypes {
  LoadParticipants = '[Context] Load Participants',
  LoadParticipantsSuccess = '[Context] Load Participants Success',
  LoadParticipantsFail = '[Context] Load Participants Fail'
}

export class LoadParticipants implements Action {
  readonly type = ContextActionTypes.LoadParticipants;
}

export class LoadParticipantsSuccess implements Action {
  readonly type = ContextActionTypes.LoadParticipantsSuccess;
  constructor(public payload: IParticipant[]) {}
}

export class LoadParticipantsFail implements Action {
  readonly type = ContextActionTypes.LoadParticipantsFail;
  constructor(public payload: any) {}
}

export type ContextActions = LoadParticipants | LoadParticipantsSuccess | LoadParticipantsFail;
