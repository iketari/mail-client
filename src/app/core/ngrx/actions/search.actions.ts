import { Action } from '@ngrx/store';
import { IEmail } from '../../../shared/models/message';
import { ISearchQuery, ISearchResponse, IParticipant } from '../../../shared/models/search';

export enum SearchActionTypes {
  LoadSearchResults = '[Search] Load Results',
  LoadSearchResultsSuccess = '[Search] Load Results Success',
  LoadSearchResultsFail = '[Search] Load Results Fail',
  SelectSearchResult = '[Search] Select Result',
  ChangeSearchParticipantsParams = '[Search] Change Participants Params',
  ChangeSearchDatesParams = '[Search] Change Date Params',
  ChangeSearchTextQuery = '[Search] Change Text Query'
}

export class LoadSearchResults implements Action {
  readonly type = SearchActionTypes.LoadSearchResults;
  constructor(public payload: { page: number; params?: Partial<ISearchQuery> }) {}
}

export class LoadSearchResultsSuccess implements Action {
  readonly type = SearchActionTypes.LoadSearchResultsSuccess;
  constructor(public payload: ISearchResponse<IEmail>) {}
}

export class LoadSearchResultsFail implements Action {
  readonly type = SearchActionTypes.LoadSearchResultsFail;
  constructor(public payload: any) {}
}

export class SelectResult implements Action {
  readonly type = SearchActionTypes.SelectSearchResult;
  constructor(public payload: { id: string }) {}
}

export class ChangeSearchParticipantsParams implements Action {
  readonly type = SearchActionTypes.ChangeSearchParticipantsParams;
  constructor(public payload: { from?: IParticipant; to?: IParticipant[] }) {}
}

export class ChangeSearchDatesParams implements Action {
  readonly type = SearchActionTypes.ChangeSearchDatesParams;
  constructor(public payload: { date_from?: Date; date_to?: Date }) {}
}

export class ChangeSearchTextQuery implements Action {
  readonly type = SearchActionTypes.ChangeSearchTextQuery;
  constructor(public payload: { query: string }) {}
}

export type SearchActions =
  | LoadSearchResults
  | LoadSearchResultsSuccess
  | LoadSearchResultsFail
  | SelectResult
  | ChangeSearchParticipantsParams
  | ChangeSearchDatesParams
  | ChangeSearchTextQuery;
