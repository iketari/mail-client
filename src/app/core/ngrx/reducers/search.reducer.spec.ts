import { reducer, initialState, State } from './search.reducer';
import { LoadSearchResults, ChangeSearchTextQuery } from '../actions/search.actions';
import { IThreadSearchResult } from '../../../shared/models/search';
import { IEmail } from '../../../shared/models/message';

describe('[Core] Search Reducer', () => {
  describe('LoadSearchResults', () => {
    it('should change state to loading', () => {
      const actual = reducer(
        initialState,
        new LoadSearchResults({
          page: 1
        })
      );

      expect(actual).toMatchSnapshot();
    });
  });

  describe('ChangeSearchTextQuery', () => {
    it('should chnage search line in the query', () => {
      const state: State = {
        selected: null,
        entities: {
          id1: <IThreadSearchResult<IEmail>>{ id: 'id1', subject: 'Hello world!' }
        },
        ids: ['id1'],
        loading: false,
        total: 1,
        page: 1,
        limit: 10,
        searchQuery: {
          from: null,
          to: [],
          date_from: null,
          date_to: null,
          query: null
        }
      };

      const actual = reducer(
        state,
        new ChangeSearchTextQuery({
          query: 'Hello world'
        })
      );

      expect(actual).toMatchSnapshot();
    });
  });
});
