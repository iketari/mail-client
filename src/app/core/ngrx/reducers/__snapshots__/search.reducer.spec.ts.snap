// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Core] Search Reducer ChangeSearchTextQuery should chnage search line in the query 1`] = `
Object {
  "entities": Object {
    "id1": Object {
      "id": "id1",
      "subject": "Hello world!",
    },
  },
  "ids": Array [
    "id1",
  ],
  "limit": 10,
  "loading": true,
  "page": 1,
  "searchQuery": Object {
    "date_from": null,
    "date_to": null,
    "from": null,
    "query": "Hello world",
    "to": Array [],
  },
  "selected": null,
  "total": 1,
}
`;

exports[`[Core] Search Reducer LoadSearchResults should change state to loading 1`] = `
Object {
  "entities": Object {},
  "ids": Array [],
  "limit": 10,
  "loading": true,
  "page": 0,
  "searchQuery": Object {
    "date_from": null,
    "date_to": null,
    "from": null,
    "query": null,
    "to": Array [],
  },
  "selected": null,
  "total": 0,
}
`;
